ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/ConflictCheckingLedgerBridge.scala

```diff
@@ -0,0 +1,623 @@
+  }
+
+  private def conflictCheckWithInTransit(
+      sequencerQueueState: SequencerState,
+      keyInputs: KeyInputs,
+      inputContracts: Set[ContractId],
+      transaction: Submission.Transaction,
+  ): Either[SoxRejection, Submission.Transaction] = {
+    val updatedKeys = sequencerQueueState.keyState
+    val archives = sequencerQueueState.consumedContractsState
+
+    keyInputs
+      .foldLeft[Either[SoxRejection, Unit]](Right(())) {
+        case (Right(_), (key, KeyCreate)) =>
+          updatedKeys.get(key) match {
+            case Some((None, _)) | None => Right(())
+            case Some((Some(_), _)) => Left(DuplicateKey(key)(transaction))
+          }
+        case (Right(_), (key, NegativeKeyLookup)) =>
+          updatedKeys.get(key) match {
+            case Some((None, _)) | None => Right(())
+            case Some((Some(actual), _)) =>
+              Left(InconsistentContractKey(None, Some(actual))(transaction))
+          }
+        case (Right(_), (key, KeyActive(cid))) =>
+          updatedKeys.get(key) match {
+            case Some((Some(`cid`), _)) | None => Right(())
+            case Some((other, _)) => Left(InconsistentContractKey(other, Some(cid))(transaction))
+          }
+        case (left, _) => left
+      }
+      .flatMap { _ =>
+        val alreadyArchived = inputContracts.intersect(archives.keySet)
+        if (alreadyArchived.nonEmpty) Left(UnknownContracts(alreadyArchived)(transaction))
+        else Right(())
+      }
+      .map(_ => transaction)
+  }
+
+  private def toRejection(rejection: SoxRejection) =
+    Update.CommandRejected(
+      recordTime = Timestamp.now(),
+      completionInfo = rejection.originalTx.submitterInfo.toCompletionInfo,
+      reasonTemplate = FinalReason(rejection.toStatus),
+    )
+
+  private def conflictCheckWithCommitted(
+      transaction: Submission.Transaction,
+      inputContracts: Set[ContractId],
+      keyInputs: KeyInputs,
+  ): Future[Either[SoxRejection, Unit]] =
+    Timed.future(
+      metrics.daml.SoX.conflictCheckWithCommitted,
+      checkTimeModel(transaction) match {
+        case Left(value) => Future(Left(value))
+        case Right(_) =>
+          validateCausalMonotonicity(
+            transaction,
+            inputContracts,
+            transaction.transactionMeta.ledgerEffectiveTime,
+            transaction.blindingInfo.divulgence.keySet,
+          ).flatMap {
+            case Right(_) => validatePartyAllocation(transaction)
+            case rejection => Future.successful(rejection)
+          }.flatMap {
+            case Right(_) => validateKeyUsages(transaction, keyInputs)
+            case rejection => Future.successful(rejection)
+          }
+      },
+    )
+
+  private def validatePartyAllocation(
+      transaction: Submission.Transaction
+  ): Future[Either[SoxRejection, Unit]] = {
+    val informees = transaction.transaction.informees
+    indexService
+      .getParties(informees.toSeq)
```

Comment:
seems very expensive/redundant.
how about having a party view in the SoX implementation instead, which can be precisely updated at the sequential state thus this check can be moved also there?